test/integration/consul-container/test/ratelimit/ratelimit_test.go

```diff
@@ -0,0 +1,248 @@
+			cmd:         `-hcl=limits { request_limits { mode = "enforcing" read_rate = 0 write_rate = 0 }}`,
+			operations: []operation{
+				{
+					action:            putKV,
+					expectedErrorMsg:  nonRetryableErrorMsg,
+					expectExceededLog: true,
+					expectMetric:      true,
+				},
+				{
+					action:            getKV,
+					expectedErrorMsg:  retryableErrorMsg,
+					expectExceededLog: true,
+					expectMetric:      true,
+				},
+			},
+		}}
+
+	for _, tc := range testCases {
+		t.Run(tc.description, func(t *testing.T) {
+			logConsumer := &TestLogConsumer{}
+			cluster := createCluster(t, tc.cmd, logConsumer)
+			defer terminate(t, cluster)
+
+			client, err := cluster.GetClient(nil, true)
+			require.NoError(t, err)
+
+			// perform actions and validate returned errors to client
+			for _, op := range tc.operations {
+				err = op.action.function(client)
+				if len(op.expectedErrorMsg) > 0 {
+					require.Error(t, err)
+					require.Equal(t, op.expectedErrorMsg, err.Error())
+				} else {
+					require.NoError(t, err)
+				}
+			}
+
+			// validate logs and metrics
+			// doing this in a separate loop so we can perform actions, allow metrics
+			// and logs to collect and then assert on each.
+			for _, op := range tc.operations {
+				timer := &retry.Timer{Timeout: 10 * time.Second, Wait: 500 * time.Millisecond}
+				retry.RunWith(timer, t, func(r *retry.R) {
+					// validate metrics
+					metricsInfo, err := client.Agent().Metrics()
+					// TODO(NET-1978): currently returns NaN error
+					//			require.NoError(t, err)
+					if metricsInfo != nil && err == nil {
+						if op.expectMetric {
+							checkForMetric(r, metricsInfo, op.action.rateLimitOperation, op.action.rateLimitType)
+						}
+					}
+
+					// validate logs
+					// putting this last as there are cases where logs
+					// were not present in consumer when assertion was made.
+					checkLogsForMessage(r, logConsumer.Msgs,
+						fmt.Sprintf("[WARN]  agent.server.rpc-rate-limit: RPC exceeded allowed rate limit: rpc=%s", op.action.rateLimitOperation),
+						op.action.rateLimitOperation, "exceeded", op.expectExceededLog)
+
+				})
+			}
+		})
+	}
+}
+
+func checkForMetric(t *retry.R, metricsInfo *api.MetricsInfo, operationName string, expectedLimitType string) {
+	for _, counter := range metricsInfo.Counters {
+		if counter.Name == "consul.rate.limit" {
+			operation, ok := counter.Labels["op"]
+			require.True(t, ok)
+
+			limitType, ok := counter.Labels["limit_type"]
+			require.True(t, ok)
+
+			mode, ok := counter.Labels["mode"]
+			require.True(t, ok)
+
+			if operation == operationName {
+				require.Equal(t, 2, counter.Count)
+				require.Equal(t, expectedLimitType, limitType)
+				require.Equal(t, "disabled", mode)
+			}
+		}
+	}
+
+}
+
+func checkLogsForMessage(t *retry.R, logs []string, msg string, operationName string, logType string, logShouldExist bool) {
+	found := false
+	for _, log := range logs {
+		if strings.Contains(log, msg) {
+			found = true
```

Comment:
Optional: it doesn't really matter because `logs` isn't huge, but once you've found `msg` you can break the loop.
 
```suggestion
			found = true
			break
```